cmd/testing/pubsub/main.go

```diff
@@ -0,0 +1,136 @@
+// Pubsub is a simple manual test for ordered pubsub.
+package main
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+	"os/signal"
+	"time"
+
+	"cloud.google.com/go/pubsub"
+	"github.com/birdie-ai/golibs/event"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// Event is the test event data.
+type Event struct {
+	PartitionID string
+	Count       int
+}
+
+func main() {
+	usage := func() {
+		log.Fatalf("usage: %q [subscriber|publisher]", os.Args[0])
+	}
+	if len(os.Args) <= 1 {
+		usage()
+	}
+	projectID := os.Getenv("GOOGLE_PROJECT")
+	if projectID == "" {
+		log.Fatal("missing env var: GOOGLE_PROJECT")
+	}
+	topicName := os.Getenv("TOPIC_NAME")
+	if topicName == "" {
+		log.Fatal("missing env var: TOPIC_NAME")
+	}
+
+	ctx, cancel := signal.NotifyContext(context.Background(), os.Interrupt)
+	defer cancel()
+
+	switch os.Args[1] {
+	case "subscriber":
+		subscriber(ctx, projectID, topicName)
+	case "publisher":
+		publisher(ctx, projectID, topicName)
+	default:
+		usage()
+	}
+}
+
+const (
+	eventName       = "golibs-test"
+	totalPartitions = 10
+	totalEvents     = 100
+)
+
+func subscriber(ctx context.Context, projectID, topicName string) {
+	client, err := pubsub.NewClient(ctx, projectID)
+	panicerr(err)
+	defer func() {
+		_ = client.Close()
+	}()
+
+	_, err = client.CreateSubscription(ctx, topicName, pubsub.SubscriptionConfig{
+		Topic:                 client.Topic(topicName),
+		ExpirationPolicy:      24 * time.Hour,
+		EnableMessageOrdering: true,
+	})
+	if err != nil && status.Code(err) != codes.AlreadyExists {
+		log.Fatalf("creating subscription: %v", err)
+	}
+
+	log.Printf("starting handler with concurrency=%d", totalPartitions)
+	sub, err := event.NewOrderedGoogleSub[Event](ctx, projectID, topicName, eventName, totalPartitions)
+	panicerr(err)
+
+	err = sub.Serve(ctx, func(_ context.Context, event Event) error {
+		fmt.Printf("partition %q: count %d\n", event.PartitionID, event.Count)
+		time.Sleep(5 * time.Second)
```

Comment:
is this needed? I ran before checking the code and I thought it was unusable slow haha

Reply:
well define "needed", I wanted to be able to notice that the 10 goroutines received the events concurrently, the sleep allowed me to confirm that (10 prints happen at once), but I can remove it now ;-)

Reply:
yeah, good to keep it.